packages/access-token-jwt/src/token-verifier.ts

```diff
@@ -0,0 +1,344 @@
+  const method = requestOptions?.method;
+  const query = requestOptions?.query;
+  const body = requestOptions?.body;
+  const isUrlEncoded = requestOptions?.isUrlEncoded ?? false;
+  const authScheme = getAuthScheme(headers)?.toLowerCase();
+  // Extract DPoP options from the provided options or use defaults
+  const {
+    dpop: {
+      enabled: dpopEnabled = DEFAULT_DPOP_ENABLED,
+      required: dpopRequired = DEFAULT_DPOP_REQUIRED,
+      iatOffset = DEFAULT_IAT_OFFSET,
+      iatLeeway = DEFAULT_IAT_LEEWAY,
+    } = {},
+  } = options;
+  let isTokenFromHeader = false;
+
+  /**
+   * Determines whether DPoP validation is required for a given request context.
+   *
+   * Validation is triggered if:
+   * - DPoP is enabled, AND (
+   *   - DPoP is required by configuration
+   *   - OR the Authorization scheme is 'DPoP'
+   *   - OR the token contains a `cnf.jkt` claim (token is bound)
+   *   - OR the request contains a DPoP header
+   * )
+   *
+   * @method shouldVerifyDPoP
+   * @param options - Validation context
+   * @returns `true` if DPoP validation should be enforced
+   */
+  function shouldVerifyDPoP(accessTokenClaims: DPoPJWTPayload): boolean {
+    if (!dpopEnabled) {
+      return false;
+    }
+
+    const hasDPoPHeader = 'dpop' in headers;
+    const isDPoPScheme = authScheme === 'dpop';
+    const hasBoundToken = 'cnf' in accessTokenClaims;
+
+    return dpopRequired || isDPoPScheme || hasBoundToken || hasDPoPHeader;
+  }
+
+  function getToken(): TokenInfo {
+    const TOKEN_RE = /^(Bearer|DPoP) (.+)$/i;
+
+    const auth = headers.authorization;
+    const match = typeof auth === 'string' && auth.match(TOKEN_RE);
+    const fromHeader = match && match[2];
+
+    const locations: TokenInfo[] = [];
+    if (fromHeader) {
+      locations.push({ location: 'header', jwt: fromHeader });
+    }
+
+    if (typeof query?.access_token === 'string') {
+      locations.push({ location: 'query', jwt: query.access_token });
+    }
+
+    if (typeof body?.access_token === 'string' && isUrlEncoded) {
+      locations.push({ location: 'body', jwt: body.access_token });
+    }
+
+    if (locations.length === 0) throw new UnauthorizedError();
+    if (locations.length > 1)
+      throw new InvalidRequestError(
+        'More than one method used for authentication'
+      );
+    return locations[0];
+  }
+
+  /**
+   * Validates the Authorization scheme based on DPoP configuration.
+   *
+   * - If DPoP is enabled and required, expect 'DPoP' scheme.
+   * - If DPoP is enabled but not required, allow 'DPoP' or 'Bearer'.
+   * - If DPoP is disabled, only allow 'Bearer'.
+   *
+   * @method verify
+   * @throws InvalidRequestError if the scheme is invalid
+   * @throws UnauthorizedError if the scheme is not allowed
+   */
+  async function verify(): Promise<VerifyJwtResult> {
+    // Extract the token from the request headers, query, or body.
+    const { jwt, location } = getToken();
+    isTokenFromHeader = location === 'header';
```

Comment:
Shouldn't this be used to effectively skip DPoP if the token came via query parameter or request body?